nslsii/temperature_controllers.py

```diff
@@ -0,0 +1,80 @@
+from ophyd import DeviceStatus, Device, Component as Cpt, EpicsSignal, Signal
+
+
+class Eurotherm(Device):
+    '''This class is used for integrating with Eurotherm controllers.
+
+    This is used for Eurotherm controllers and is designed to ensure that the
+    set returns 'done' status only after the temperature has reached
+    equilibrium at the required value not when it first reaches the required
+    value. This is done via the attributes `self.equilibrium_time` and
+    `self.tolerance`. It only returns `done` if `self.readback` remains within
+    `self.tolerance` of `self.setpoint` over `self.equilibrium_time`. A third
+    attribute, `self.timeout`, is used to determeine the maximum time to wait
+    for equilibrium. If it takes longer than this it raises a TimeoutError.
+
+    Parameters
+    ----------
+    pv_prefix : str.
+        The PV prefix that is common to the readback and setpoint PV's.
+    '''
+
+    def __init__(self, pv_prefix, **kwargs):
+        super().__init__(pv_prefix, **kwargs)
+        self._set_lock = False
+
+    # Setup some new signals required for the moving indicator logic
+    equilibrium_time = Cpt(Signal, value=5)
+    timeout = Cpt(Signal, value=500)
+    tolerance = Cpt(Signal, value=1)
+
+    # Add the readback and setpoint components
+    setpoint = Cpt(EpicsSignal, 'SP')
+    readback = Cpt(EpicsSignal, 'I')
+
+    # define the new set method with the new moving indicator
+    def set(self, value):
+        # check that a set is not in progress, and if not set the lock.
+        if self._set_lock:
+            raise Exception('attempting to set {}'.format(self.name) +
+                            'while a set is in progress'.format(self.name))
+        self._set_lock = True
+
+        # define some required values
+        set_value = value
+        status = DeviceStatus(self, timeout=self.timeout.get())
+
+        initial_timestamp = None
+
+        # grab these values here to avoidmutliple calls.
+        equilibrium_time = self.equilibrium_time.get()
+        tolerance = self.tolerance.get()
+
+        # set up the done moving indicator logic
+        def status_indicator(value, timestamp, **kwargs):
+            nonlocal initial_timestamp
+            if abs(value - set_value) < tolerance:
+                if initial_timestamp:
+                    if (timestamp - initial_timestamp) > equilibrium_time:
```

Comment:
This assumes that we will happen to get a new reading soon enough that we aren’t left waiting to long after the equilibrium time has passed. It could hang forever. Should instead start a timer that will finish the status object unless interrupted by seeing an incompatible reading.

Reply:
Let me make sure I understand correctly what this supposed to do:
1. Set the setpoint and subscribe to updates from the readback.
3. As soon as we see a reading within the tolerance of the setpoint, start a timer (effectively).
4. If we see a reading outside the tolerance, restart the timer.
5. If we are another reading inside the tolerance and if the equilibration time has passed, we are done.

The bug I see is that that last reading could be a long time in coming, and we might be well beyond the equilibration time by then. At best we waste little bits of experiment time, at worst something times out and we get a false failure. I suggest reworking 5 so that we are done immediately when the equilibration time expires, not waiting for the first reading after that time.

Reply:
Let me make sure I understand correctly what this supposed to do:
1. Set the setpoint and subscribe to updates from the readback.
3. As soon as we see a reading within the tolerance of the setpoint, start a timer (effectively).
4. If we see a reading outside the tolerance, restart the timer.
5. If we are another reading inside the tolerance and if the equilibration time has passed, we are done.

The bug I see is that that last reading could be a long time in coming, and we might be well beyond the equilibration time by then. At best we waste little bits of experiment time, at worst something times out and we get a false failure. I suggest reworking 5 so that we are done immediately when the equilibration time expires, not waiting for the first reading after that time.

Reply:
Let me make sure I understand correctly what this supposed to do:
1. Set the setpoint and subscribe to updates from the readback.
3. As soon as we see a reading within the tolerance of the setpoint, start a timer (effectively).
4. If we see a reading outside the tolerance, restart the timer.
5. If we are another reading inside the tolerance and if the equilibration time has passed, we are done.

The bug I see is that that last reading could be a long time in coming, and we might be well beyond the equilibration time by then. At best we waste little bits of experiment time, at worst something times out and we get a false failure. I suggest reworking 5 so that we are done immediately when the equilibration time expires, not waiting for the first reading after that time.

Reply:
So that was the original structure that I had in the PR, and after comments from @tacaswell above, and 1-1 discussions with him I changed it to this structure.
To elaborate:
What I initially had was a `sleep(equilibrium_time)` after the first time we were in-band and then I checked after that time and exited if we where still in-band. This did not work as the second `self.readback.get()` returned the same value before and after the `sleep`. I discussed this issue with @tacaswell and he suggested the current implementation as the solution. 

If someone has a better solution than using `sleep` to allow for the wait I am happy to look at implementing it, but we have tried using `sleep` and that did not work.

Reply:
The eurothrem updates on a clock (which is a PV we can control).

You can not sleep in this callback because it blocks libca processing any other updates.

Reply:
That was the key fact missing from this discussion. In that case we waste at most one click cycle. Downgrading from “show-stopping” to “not optimal.”

I was suggesting sleeping in anew thread,perhaps via a Timer object, exactly as we do for settle_time. 👎on inventing a new name for what is the temperature controller equivalent of settle time because that breaks hardware generality a little but it’s not a hill I’ll die on.

Reply:
That was the key fact missing from this discussion. In that case we waste at most one click cycle. Downgrading from “show-stopping” to “not optimal.”

I was suggesting sleeping in anew thread,perhaps via a Timer object, exactly as we do for settle_time. 👎on inventing a new name for what is the temperature controller equivalent of settle time because that breaks hardware generality a little but it’s not a hill I’ll die on.

Reply:
That was the key fact missing from this discussion. In that case we waste at most one click cycle. Downgrading from “show-stopping” to “not optimal.”

I was suggesting sleeping in anew thread,perhaps via a Timer object, exactly as we do for settle_time. 👎on inventing a new name for what is the temperature controller equivalent of settle time because that breaks hardware generality a little but it’s not a hill I’ll die on.

Reply:
Thanks for the feedback @danielballan, hopefully with time I will get better at specifying the issues in a way that is clearer to you.

In regard to the name `equilibrium_time` this was also raised by  @tacaswell .

> **tacaswell 16 days ago Member**
> We already have a 'settle_time' on positioners, we probably should re-use that name here.

>**awalter-bnl 16 days ago Member**
>This is different from a settle time. A settle time is how long to wait, after motion is complete, before >moving on. This is how long to wait before rechecking that the temperature is within the tolerance >range, and hence the temperature is 'stable'. I think it makes sense to use a different term here as >they are different values.

So far I haven't heard back from @tacaswell regarding if he thinks my argument here makes sense, I prefer a different name because I am not convinced it is the temperature controller equivalent of `settle_time`, but will change it if this argument does not convince others.